logstash-core/lib/logstash/pipelines_registry.rb

```diff
@@ -0,0 +1,145 @@
+# encoding: utf-8
+
+module LogStash
+  class PipelineState
+    attr_reader :pipeline_id, :pipeline
+
+    def initialize(pipeline_id, pipeline)
+      @pipeline_id = pipeline_id
+      @pipeline = pipeline
+      @reloading = Concurrent::AtomicBoolean.new(false)
+    end
+
+    def terminated?
+      # a reloading pipeline is never considered terminated
+      @reloading.false? && @pipeline.finished_execution?
+    end
+
+    def set_reloading(is_reloading)
+      @reloading.value = is_reloading
+    end
+
+    def set_pipeline(pipeline)
+      @pipeline = pipeline
+    end
+  end
+
+  class PipelinesRegistry
+    attr_reader :states
+    include LogStash::Util::Loggable
+
+    def initialize
+      # we leverage the semantic of the Java ConcurrentHashMap for the
+      # compute() method which is atomic; calling compute() concurrently
+      # will block until the other compute finishes so no mutex is necessary
+      # for synchronizing compute calls
+      @states = java.util.concurrent.ConcurrentHashMap.new
+    end
+
+    def create_pipeline(pipeline_id, pipeline, &create_block)
+      success = false
+
+      @states.compute(pipeline_id) do |_, state|
```

Comment:
Well, we're leveraging the `ConcurrentHashMap` concurrency features and in particular the `compute` method semantics which allows us to avoid using a global lock which was the original purpose of using a `ConcurrentHashMap`. In all fairness, the only place the state object is leaked outside the `PipelineRegistry` is in the `reload` method and I think we could avoid that, I'll check. Other then that I think it is relatively straight forward. I'll go ahead and see if we can avoir leaking the state object and will also add comments to explain the logic. I did add a comment in  `PipelineRegistry#initialize` method about the `compute` method usage.